rfcs/001-mid-level-file-api.md

```diff
@@ -0,0 +1,187 @@
+# Mid-Level File Api
+
+## Summary
+
+This is a mid-level API wrapper around the various File related JavaScript apis. The possibility for a higher level API is left open.
+
+The mid-level file API aims to implement File I/O on top of the raw JS apis found here: https://w3c.github.io/FileAPI/ which includes listing files, creating files, and reading files.
+
+## The API
+
+First we have `FileList` which is simply a list of `File`s. This is an opaque struct that offers a way to iterate over `Files` and get individual files by index:
+
+```rust
+#[derive(Debug, Clone)]
+struct FileList { ... }
+
+impl FileList {
+  fn get(index: usize) -> Option<File> { ... }
+
+  fn len(&self) -> usize { ... }
+
+  fn iter(&self) -> FileListIter { ... }
+
+  fn as_raw(&self) -> &web_sys::FileList { ... }
+}
+
+impl AsRef<web_sys::FileList> for FileList { ... }
+impl From<web_sys::FileList> for FileList { ... }
+impl Index<usize> for FileList {
+    type Output = File;
+    fn index(&self, index: usize) -> &Self::Output { ... }
+}
+```
+
+There is no `FileList::new` since creating a raw `web_sys::FileList` is not possible without going through a `web_sys::HtmlInputElement`.
+
+Next we have a `blob` module that includes the trait `BlobLike`
+```rust
+trait BlobLike {
+    fn size(&self) -> u64 { ... }
+
+    // the mime crate and mimes from blobs both conform to rfc6838
+    #[cfg(feature = "mime")]
+    fn mime_type(&self) -> Result<mime::Mime, mime::FromStrError> { ... }
+
+    fn raw_mime_type(&self) -> String { ... }
+
+    fn as_raw(&self) -> &web_sys::Blob;
+}
+```
+There are two structs that implement this trait: `Blob` and `File`.
+
+```rust
+#[derive(Debug, Clone)]
+struct Blob { ... }
+
+impl Blob {
+  fn new<T>(content: T) -> Blob
+    where
+        T: std::convert::Into<BlobContents> // We'll look at BlobContents below
+    { ... }
+
+  fn new_with_options<T>(content: T, mime_type: Option<String>) -> Blob
+    where
+        T: std::convert::Into<BlobContents>
+    { ... }
+
+  fn slice(&self, start: usize, end: usize) -> Blob { ... }
```

Comment:
As above, this needs to be `f64` or `u32`

Reply:
Not sure what to do here. `u64` is a possibility for `start` and `end` if we just use `min(start,  f64::MAX)`.

Reply:
After @fitzgen 's explanation, I'm now on board with `u64` (still not `usize` though).

However, I don't think it should silently saturate, it should panic if it goes out of range.

Also the maximum needs to be `9007199254740991`, not `f64::MAX`.